---
title: "CLL scRNAseq Project Update"
author: "Brad Blaser"
date: "2/18/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
	echo = TRUE,
	message = FALSE,
	warning = FALSE
)
```

## Introduction

Priya and I analyzed the following samples for gene expression:

```{r}
read_excel("~/network/X/Labs/Blaser/single_cell/cll_project/analysis_configs.xlsx",sheet = "metadata") %>% 
  select(patient,
         ID,
         btk_clone_vaf_pct,
         relapse_vaf_pct,
         ck = complex_karyotype,
         del17p, 
         del11q,
         del13p,
         tri12)
```

We have not yet gotten into the BCR or TCR data.

Basic sequence metrics for all of the samples looked good after the necessary resequencing:

```{r}
print(summarized_sequencing_metrics, n = Inf)
```

## Data Analysis:  QC

We used a standard qc approach to remove low quality cells with high percent of mitochondrial reads or low number of detected genes.  We also use a computational approach to model and predict cell doublets.  After filtering out doublets and poor quality cells, we retained the following cells per sample:


```{r}
as_tibble(colData(cds_aligned)) %>% 
  group_by(pt, timepoint) %>% 
  summarise(n = n()) %>% 
  ungroup() %>% 
  mutate(cumulative_total = cumsum(n)) %>%
  print(n = Inf)
```

## Data Analysis:  Dimensionality Reduction and Clustering

The next task is usually to perform dimensionality reduction, cluster cells and identify clusters.  

Using UMAP dimensionality reduction to visualize global gene expression, we see that the dataset is highly heterogeneous:

```{r dev='png',fig.align='center',fig.width=7.5,fig.height=4,dpi=300}
umap_unaligned
```

There are too many colors here to tell apart, but what you can probably tell is that there are two clusters with many colors in them and many clusters with one or two colors in them.  The multicolor clusters are T cells and Monocytes and the rest are mostly B cells. 

Rather than go through each cluster individually, we can use batch correction techniques to align the similar clusters:

```{r dev='png',fig.align='center',fig.width=7.5,fig.height=4,dpi=300}
umap_sample
```

Here is the entire data set plotted by patient and timepoint (timepoints 1, 2 and 3):

```{r dev='png',fig.align='center',fig.width=5.5,fig.height=10.0,dpi=300}
umap_allpts_alltimes
```

Generally speaking, if we think the sample to sample variability is biological, as is the case here, we would prefer only to use the aligned dataset for clustering and categorizing similar cell types and to use the unaligned data for things like differential gene expression.  This is the approach we are taking here.  If however, we felt like there was technical variation related to sample preparation for example, then we would align according to that variable first and use the aligned data for differential gene expression.  I just point that out to illustrate the approach.

Next we can use 4 different algorithms and approaches to identify what cells are what.  The first 3 I will show use various unsupervised classification algorithms to identify coarse, medium and fine resolution clusters:

```{r dev='png',fig.align='center',fig.width=5,fig.height=4,dpi=300}
umap_partition
```

```{r dev='png',fig.align='center',fig.width=5,fig.height=4,dpi=300}
umap_leiden
```

```{r dev='png',fig.align='center',fig.width=5,fig.height=4,dpi=300}
umap_louvain
```

Each of these can be fine-tuned to get more or less resolution, but these were run with default settings.  At this stage of the game, the coarse partition clusters are more useful.  Later we can use the leiden or louvain clusters to drill down into populations if we want.

At this point we would typically get a list of specific genes for each partition and then assign the clusters by inspection.

Instead, we can use the fourth method to cluster and automatically assign cluster IDs according to a highly annotated PBMC dataset.  This comes from the Seurat v4 package and uses a technique called label transfer.  Our data is mapped to a set of a couple hundred thousand cells that were run using 10X tecnnology but with the addition of a giant panel of cell surface antibodies (cite-seq).  

The label transfer approach is nice because it does the work of assigning cluster/celltype IDs for you.  You get the benefit of somebody else having run a giant cite-seq experiment without having to buy $100K in antibiodies (not exaggerating).  Also, people in the community really like Seurat.

The disadvantage is that it is a supervised method built on a normal PBMC reference.  If your celltype is highly abnormal, then Seurat will give it a label according to the nearest normal counterpart.  Most AML blasts for example it will call CD14 monocytes, which of course is incorrect.  The other issue is that the way Seurat does the clustering, your cells will end up mapped onto the reference dataset UMAP coordinates like this:

```{r dev='png',fig.align='center',fig.width=5,fig.height=4,dpi=300}
umap_seurat_coords_l1
```

```{r dev='png',fig.align='center',fig.width=5,fig.height=4,dpi=300}
umap_seurat_coords_l2
```

These are just two different levels of clustering granularity.  You can see how it is helpful, but the problem is that all of our B cells are stacked up on top of each other.  Any dataset you map with this technique will have the same x-y pattern.  It is like everything is silkscreened with the same stencil, but with different amounts of paint in different areas.  So this is not great by itself.

What we can do instead is take these cell classifications, and plot the cells according to our original UMAP coordinates:
Here are the data with the automated celltype labels from seurat:

```{r dev='png',fig.align='center',fig.width=5,fig.height=4,dpi=300}
umap_seurat_l1
```

```{r dev='png',fig.align='center',fig.width=5,fig.height=4,dpi=300}
umap_seurat_l2
```

As you can see, the problem here is that the clusters get dispersed to some degree when you change coordinates.  For example, the other cluster from umap_seurat_l1:

```{r dev='png',fig.align='center',fig.width=5,fig.height=4,dpi=300}
umap_seurat_l1_other
```

Overall in this dataset, I prefer to use the seurat l1 (coarse) clustering at this point.  We can debate this approach and revise.  The other cells here are highlighed but end up being a small part of the overall dataset.  The rest looks pretty good.

This heatmap shows cluster membership according to timepoint:

```{r dev='png',fig.align='center',fig.width=5,fig.height=4,dpi=300}
plot_grid(cluster_membership_heatmap)
```

It looks like CD8s go up from timepoint 2 to 3 and monocytes and CD4s go down.

## Data Analysis:  Differential Gene Expression by Pseudobulk Analysis

The most rigorous way to determine differential gene expression is to group the cells you want, calculate an aggregete gene count matrix for all of those cells according to variables defining experimental classes (timepoint) and replicates (samples).  We use DESeq2 to do the DGE calculations.  This is called pseudobulk analysis.  It treats each sample as a biological replicate, rather than each cell.  This generally underestimates the differences because it doesn't give you any credit for repeatedly detecting genes in multiple cells.  This should improve your confidence around the detection of a particular gene in a certain cellular state. However I would say it is the most reviewer-proof method out there that I know of.

I performed pseudobulk analysis on the B cell cluster, treating each sample as a replicate and making all pairwise comparisons between timepoints 1, 2 and 3.  

There were only 16 significantly different genes between timepoints 1 and 2 and none between 1 and 3 or 2 and 3:

```{r}
pseudobulk_res[[1]][[2]] %>% filter(padj<0.05)
```

In this table, a positive log2 fold change indicates up in timepoint 2 compared to timepoint 1

## Data Analysis:  DGE using Monocle Top Markers

There are less stringent DGE methods we can use to determine DGE.  The most useful at this point is a function in Monocle called Top Markers.  Here we select B cells and tell the computer to calculate N = 50 top markers for the B cells at each timepoint.  Monocle uses 2 calculations to determine the markers:  it calculates the Jensen-Shannon Divergence and then performs logistic regression based on the variables we input.

If you are using the project repository, these data are at "data_out/timepoint_top_markers.csv".  I am also attaching to this message.

## Data Analysis:  Gene Modules

Another powerful way to extract meaning from this dataset is to use gene modules.  These are groups of co-regulated genes that go up and down in a similar way across the dataset.  We calculate the gene modules using the UMAP and clustering algorithms, just like we do for cells, but in an inverse way if that makes sense.  The gene modules are defined in a way that is agnostic to any cellular metadata, so once we have defined the modules, we can ask the computer to plot aggregate expression of the modules across arbitrary groups of cells.  Gene modules may have dozens or thousands of genes in them.

Here is the gene module expression across cellular clusters and timepoints:

```{r dev='png',fig.align='center',fig.width=7.5,fig.height=5,dpi=300}
plot_grid(gene_module_heatmap)
```
The idea here, is you want to look at how the cell and timepoint catgories cluster within the dendrograms and then use the heatmap to determine which modules drive the clustering.  You can see that generally, B, T and monos cluster together.  The other cells are most similar to B cells.  Module 1 and 7 go down from timepoint 1 to 3.  Module 9 goes up.  Module 5 is neutral in B cells but high in other cells at timepoints 2 and 3.  

All of the module genes are in the file "data_out/gene_module_df_anno.csv" if you want to look.

The next thing we do is use GO term analysis to get some biologcial meaning out of these gene lists.  There are a few steps we take to narrow down the GO term lists, but here is what we end up with (apologies for the overlapping labels; we can fix those in post):


```{r dev='png',fig.align='center',fig.width=7.5,fig.height=6,dpi=300}
module_bubble_plots$`Module 1`
```


```{r dev='png',fig.align='center',fig.width=7.5,fig.height=6,dpi=300}
module_bubble_plots$`Module 5`
```

```{r dev='png',fig.align='center',fig.width=7.5,fig.height=6,dpi=300}
module_bubble_plots$`Module 7`
```

```{r dev='png',fig.align='center',fig.width=7.5,fig.height=6,dpi=300}
module_bubble_plots$`Module 9`
```

## Summary

We have a lot of data to look through.  Now would be a great time for Priya to look at all of the gene lists and come up with some hypotheses about what it all means.  Some of these hypotheses we can probably address just with additional analysis.  

Once we have the BTK data, we can use that as an additional categorical variable for the gene modules, top markers, pseudobulk etc.  

The BCR and TCR data can be overlaid relatively quickly on these plots; I just haven't been able to get to it yet. We can look for clonal restriction in T cells, query some databases to look for pathology associated TCR sequences (usually you come up empty handed with this but you never know.), etc.  This is more interesting if we have some thoughts about T cell phenotypic changes.  At a minimum we will want to show the malignant clone with the BCR data.


